### ISSUE TITLE
Seeking sane approach to ameliorate Modifier + F-keys inconsistency

### ISSUE THREAD

**Issue by unphased** [tmux/tmux#2777]
I use Alacritty and tmux together on my main machine. What i've found is that the following are the default Shift+F-key bindings:

```
❯ cat                                                                                        
^[[1;2P^[[1;2Q^[[1;2R^[[1;2S^[[15;2~^[[17;2~^[[18;2~^[[19;2~^[[20;2~^[[21;2~^[[23;2~^[[24;2~
```
From within tmux:
```
^[[25~^[[26~^[[28~^[[29~^[[31~^[[32~^[[33~^[[34~^[[20;2~^[[21;2~^[[23;2~^[[24;2~
```
I can do further testing to confirm but don't think I need to; tmux is able to interpret e.g. <kbd>Esc</kbd>`[1;2P` as S-F1 and emitting <kbd>Esc</kbd>`[25~` to its ptys, and I more or less get that it does this not out of any particularly specific reason. This is simply what it does. We can see here https://invisible-island.net/xterm/xterm-function-keys.html that it is sensibly adhering to the vt220 binding and seemingly switching to xterm-new (whatever that is) for S-F9 and up.  I've no doubt there exist some terminal emulators that exhibit the same behavior. 

My question is merely a manifestation of simultaneously being completely bewildered about the whole thing and being a relative expert in it owing to nearly a decade now of repeated experimentation in this space. 

The immediate problem is that I cannot properly bind `<S-F7>` from vim, vim expects <kbd>Esc</kbd>`[18;2P` but is receiving <kbd>Esc</kbd>`[33~` if I run vim in tmux, so my new binding only works outside tmux. 

I have a number of options in terms of what's possible. For now the easiest is to train vim to understand tmux via `set <S-F7>=^[[33~`. 

But it is clear that both alacritty and vim are agreeing that shift F7 should be <kbd>Esc</kbd>`[18;2P`: The question, then, is should tmux do something about it? 

I'm also aware that I could tell tmux about it explicitly, via `set -g terminal-overrides "*:kf13=\\033[25~,*:kf14=\\033[26~"`, etc. I think that's gotta be my answer but I find it to be unsatisfying.

What do you think?

**Comment by nicm** [comment]
tmux follows screen and xterm here and the screen and tmux terminfo entries match what it sends. If vim does not correctly recognise them it seems like either your TERM inside tmux is incorrect or there is a bug in vim.

**Comment by unphased** [comment]
I guess part of me is surprised that tmux correctly interprets the <kbd>Esc</kbd>`[1;2P` style codes at all.

**Comment by nicm** [comment]
Why wouldn't it? It has to understand them or it couldn't translate them to the right thing.

**Comment by unphased** [comment]
Ah yes. Well thanks for reminding me. I took a look at `infocmp` and the definition for my particular $TERM is very sparse, missing the entries for those function keys. Whereas the plain `xterm` one does define them, so I'll do some testing and then give customizing it a shot. As you say, I fully expect tmux to properly interpret all of this, so it must be a bug in the other software.

**Comment by nicm** [comment]
tmux understands a bunch of stuff builtin because terminfo can be limited.

**Comment by unphased** [comment]
I changed my tmux-256color-italic.terminfo to copy xterm rather than screen, so  now it should be different, but it still exhibits the old behavior. 

```
slu@slu-5950X:~ 5950X 20.04[2] 07/14/21 13:48:21 5006
❯ cat                                                                                                                []
^[[25~^[[26~^[[28~^[[25~^C
slu@slu-5950X:~ 5950X 20.04[tmux] 07/14/21 13:48:25 5007
❯ infocmp                                                                                                     130 ↲  []
#       Reconstructed via infocmp from file: /home/slu/.terminfo/t/tmux-256color-italic
tmux-256color-italic|screen with 256 colors and italic,
        am, bce, ccc, km, mc5i, mir, msgr, npc, xenl,
        colors#0x100, cols#80, it#8, lines#24, pairs#0x10000,
        acsc=``aaffggiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz{{||}}~~,
        bel=^G, blink=\E[5m, bold=\E[1m, cbt=\E[Z, civis=\E[?25l,
        clear=\E[H\E[2J, cnorm=\E[?12l\E[?25h, cr=\r,
        csr=\E[%i%p1%d;%p2%dr, cub=\E[%p1%dD, cub1=^H,
        cud=\E[%p1%dB, cud1=\n, cuf=\E[%p1%dC, cuf1=\E[C,
        cup=\E[%i%p1%d;%p2%dH, cuu=\E[%p1%dA, cuu1=\E[A,
        cvvis=\E[?12;25h, dch=\E[%p1%dP, dch1=\E[P, dim=\E[2m,
        dl=\E[%p1%dM, dl1=\E[M, ech=\E[%p1%dX, ed=\E[J, el=\E[K,
        el1=\E[1K, flash=\E[?5h$<100/>\E[?5l, home=\E[H,
        hpa=\E[%i%p1%dG, ht=^I, hts=\EH, ich=\E[%p1%d@,
        il=\E[%p1%dL, il1=\E[L, ind=\n, indn=\E[%p1%dS,
        initc=\E]4;%p1%d;rgb\:%p2%{255}%*%{1000}%/%2.2X/%p3%{255}%*%{1000}%/%2.2X/%p4%{255}%*%{1000}%/%2.2X\E\\,
        invis=\E[8m, is2=\E[!p\E[?3;4l\E[4l\E>, kDC=\E[3;2~,
        kEND=\E[1;2F, kHOM=\E[1;2H, kIC=\E[2;2~, kLFT=\E[1;2D,
        kNXT=\E[6;2~, kPRV=\E[5;2~, kRIT=\E[1;2C, kb2=\EOE, kbs=^?,
        kcbt=\E[Z, kcub1=\EOD, kcud1=\EOB, kcuf1=\EOC, kcuu1=\EOA,
        kdch1=\E[3~, kend=\EOF, kent=\EOM, kf1=\EOP, kf10=\E[21~,
        kf11=\E[23~, kf12=\E[24~, kf13=\E[1;2P, kf14=\E[1;2Q,
        kf15=\E[1;2R, kf16=\E[1;2S, kf17=\E[15;2~, kf18=\E[17;2~,
        kf19=\E[18;2~, kf2=\EOQ, kf20=\E[19;2~, kf21=\E[20;2~,
        kf22=\E[21;2~, kf23=\E[23;2~, kf24=\E[24;2~,
        kf25=\E[1;5P, kf26=\E[1;5Q, kf27=\E[1;5R, kf28=\E[1;5S,
        kf29=\E[15;5~, kf3=\EOR, kf30=\E[17;5~, kf31=\E[18;5~,
        kf32=\E[19;5~, kf33=\E[20;5~, kf34=\E[21;5~,
        kf35=\E[23;5~, kf36=\E[24;5~, kf37=\E[1;6P, kf38=\E[1;6Q,
        kf39=\E[1;6R, kf4=\EOS, kf40=\E[1;6S, kf41=\E[15;6~,
        kf42=\E[17;6~, kf43=\E[18;6~, kf44=\E[19;6~,
        kf45=\E[20;6~, kf46=\E[21;6~, kf47=\E[23;6~,
        kf48=\E[24;6~, kf49=\E[1;3P, kf5=\E[15~, kf50=\E[1;3Q,
        kf51=\E[1;3R, kf52=\E[1;3S, kf53=\E[15;3~, kf54=\E[17;3~,
        kf55=\E[18;3~, kf56=\E[19;3~, kf57=\E[20;3~,
        kf58=\E[21;3~, kf59=\E[23;3~, kf6=\E[17~, kf60=\E[24;3~,
        kf61=\E[1;4P, kf62=\E[1;4Q, kf63=\E[1;4R, kf7=\E[18~,
        kf8=\E[19~, kf9=\E[20~, khome=\EOH, kich1=\E[2~,
        kind=\E[1;2B, kmous=\E[M, knp=\E[6~, kpp=\E[5~,
        kri=\E[1;2A, mc0=\E[i, mc4=\E[4i, mc5=\E[5i, oc=\E]104\007,
        op=\E[39;49m, rc=\E8, rev=\E[7m, ri=\EM, rin=\E[%p1%dT,
        ritm=\E[23m, rmacs=\E(B, rmam=\E[?7l,
        rmcup=\E[?1049l\E[23;0;0t, rmir=\E[4l, rmkx=\E[?1l\E>,
        rmm=\E[?1034l, rmso=\E[27m, rmul=\E[24m,
        rs1=\Ec\E]104\007, rs2=\E[!p\E[?3;4l\E[4l\E>, sc=\E7,
        setab=\E[%?%p1%{8}%<%t4%p1%d%e%p1%{16}%<%t10%p1%{8}%-%d%e48;5;%p1%d%;m,
        setaf=\E[%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e38;5;%p1%d%;m,
        sgr=%?%p9%t\E(0%e\E(B%;\E[0%?%p6%t;1%;%?%p5%t;2%;%?%p2%t;4%;%?%p1%p3%|%t;7%;%?%p4%t;5%;%?%p7%t;8%;m,
        sgr0=\E(B\E[m, sitm=\E[3m, smacs=\E(0, smam=\E[?7h,
        smcup=\E[?1049h\E[22;0;0t, smir=\E[4h, smkx=\E[?1h\E=,
        smm=\E[?1034h, smso=\E[7m, smul=\E[4m, tbc=\E[3g,
        u6=\E[%i%d;%dR, u7=\E[6n, u8=\E[?%[;0123456789]c,
        u9=\E[c, vpa=\E[%i%p1%dd,
slu@slu-5950X:~ 5950X 20.04[tmux] 07/14/21 13:48:33 5008
❯                                                                     
```

**Comment by nicm** [comment]
I don't know, best ask the vim guys. Better to get the tmux terminfo from somewhere rather than copying xterm.

**Comment by nicm** [comment]
https://gist.github.com/nicm/ea9cf3c93f22e0246ec858122d9abea1

**Comment by unphased** [comment]
I used that, configured tmux to use tmux-256color, confirmed a few things (infocmp, and that italics don't work anymore, so it's def using the new one), but it is still emitting the wrong shift-F-keys. Hmmm. 

I tested it by running `tmux -L test` to start a separate server since i still have work open in my main one.

**Comment by unphased** [comment]
My understanding is if kf13 is specified as `kf13=\E[1;2P` in terminfo then tmux should be emitting that instead of `\E[25~` when i press shift+F1.

**Comment by nicm** [comment]
tmux-256color absolutely supports italics if you set default-terminal to tmux-256color.

What do you mean emitting the wrong shift keys? tmux is emitting keys that do not match terminfo?

**Comment by nicm** [comment]
tmux will always send \E[25~, that is what tmux as a terminal sends for S-F1, TERM inside tmux tells applications inside tmux what to expect, not tmux what to send.

**Comment by unphased** [comment]
Oh. Doesnt this show then that the terminfo file you linked me isn't even self-consistent with tmux's own behavior?

**Comment by nicm** [comment]
It should match, what tmux version do you have?

**Comment by nicm** [comment]
I don't know how you are persuading tmux to send `\E[25~` because it should send `\E[1;2P` for S-F1:
~~~~
$ tmux send S-F1;cat
^[[1;2P
~~~~
Are you using an old tmux with `xterm-keys` off?

**Comment by nicm** [comment]
Oh there was a bug which was fixed in f2951bd4a560692048e68bf29a5b0a288346a7d4, are you using a tmux older than that?

**Comment by nicm** [comment]
This was fixed in 3.2a as 47af583a50db35e1fbf881845d0aa342e8fa57c8. I guess you need to update?

**Comment by unphased** [comment]
Perfect! Thanks a bunch, as always!
